doc/hare-dtm-hld.md

```diff
@@ -0,0 +1,614 @@
+hax1 -> consul: get Motr2 ha state
+hax2 -> consul: get Motr2 ha state
+hax3 -> consul: get Motr2 ha state
+    loop if Motr2 ha state is M0_NC_RECOVERING
+        hax1 -> hax1: do nothing, let Motr2 local, hax2 handle
+        hax2 -> hax2: wait for Motr2 to restart
+        hax3 -> hax3: do nothing, let Motr2 local hax2 handle
+    end
+Motr2 -> hax2: restarts, requests entrypoint
+hax2 -> Motr2: sends entrypoint reply, ha states
+hax2 -> consul: post BQ event to broadcast M0_NC_TRANSIENT
+hax2 -> consul: post BQ event to broadcast M0_NC_RECOVERING
+hax2 -> hax2: wait for dtm recovery to complete
+@enduml
+-->
+![img](recovering-process-failure.png)
+
+#### Participating process failure
+If a participant Motr process fails during ongoing DTM recovery, then Hare
+broadcasts M0_NC_FAILED notification to all the Motr processes, in this case
+the ongoing DTM recovery may complete or fail. In case of DTM recovery failure,
+Hare transitions the RECOVERING process to TRANSIENT and waits for the next
+event.
+<!--
+@startuml
+autonumber
+ha_consul -> hax1_RC: notify Motr2 failure
+ha_consul -> hax2: notify Motr2 failure
+ha_consul -> hax3: notify Motr2 failure
+hax1_RC -> consul: post a EQ event to update failvec and process ha state
+hax2 -> consul: post a EQ event to update failvec and process ha state
+hax3 -> consul: post a EQ event to update failvec and process ha state
+hax1_RC -> consul: process EQ, update failvec and Motr2 ha state, discards duplicates
+Motr2 -> hax2: restarts, requests entrypoint
+hax2 -> consul: fetches failvec for current Motr2 epoch
+hax2 -> Motr2: send entrypoint reply with ha states
+hax2 -> consul: update Motr2 ha state to M0_NC_TRANSIENT and broadcast (BQ) the same
+Motr2 -> hax2: notifies M0_CONF_HA_PROCESS_STARTING
+hax2 -> consul: updates Motr2 process status to M0_CONF_HA_PROCESS_STARTING and ha state to M0_NC_RECOVERING
+hax2 -> hax2: wait for dtm recovery to complete
+destroy Motr1
+ha_consul -> hax1_RC: notify Motr1 failure
+ha_consul -> hax2: notify Motr1 failure
+ha_consul -> hax3: notify Motr1 failure
+hax2 -> Motr2: notify Motr1 failure
+hax3 -> Motr3: notify Motr1 failure
+hax1_RC -> consul: post an EQ event to update failvec and Motr1 ha state
+hax2 -> consul: post an EQ event to update failvec and Motr1 ha state
+hax3 -> consul: post an EQ event to update failvec and Motr1 ha state
+hax1_RC -> consul: process EQ, update failvec and Motr1 ha state, discards duplicate
+hax2 -> hax2: wait for dtm recovery to complete
+Motr2 -> hax2: report dtm recovery failure or success
+    loop if dtm recovery failure
+        hax2 -> consul: reset Motr2 ha state to M0_NC_TRANSIENT
+    end
+@enduml
+-->
+![img](participant-failure.png)
+
+
+## External Interfaces
+[This section describes any external interfaces used by or exposed by Hare.]
+
+Hare uses Motr Spiel API to send DTM0 specific messages to Motr if required,
+e.g. DTM0 fops.
+
+#### Todo
+- [x] update actual Motr Spiel interfaces.
```

Comment:
I think that we also need to specify (at least, very high level) DTM API. E.g. how to initiate DTM recovery, how to make sure that DTM recovery is complete etc.

Reply:
Agree, but the information is not yet available completely. I am expecting this information from Anatoliy. I will still try to find out more on this before we start implementation.

Reply:
For instance, when I started reading this document, I used to think that DTM is a kind of long-running process and thus we can assume some API like "get DTM operation status by its ID". That's why I used the word "polling" in my previous questions. Now it seems to me that there will be no ID, and the result of DTM operation will be reported via callbacks from C land.

That's also fine. But then we need to understand what are the guarantees regarding those callbacks. Like
1. Which hax will be notified with this callback?
2. What will happen if Motr processes die and restart in the middle of DTM recovery? Will recovery proceed automatically? Will the callback be invoked anyways?

So the more clarity we can give now (at least as an assumption), the better.

Reply:
>For instance, when I started reading this document, I used to think that DTM is a kind of long-running process and thus we can assume some API like "get DTM operation status by its ID". That's why I used the word "polling" in my previous questions. Now it seems to me that there will be no ID, and the result of DTM operation will be reported via callbacks from C land.

Correct.

>Which hax will be notified with this callback?

Only the local hax (local to the recovering Motr process) receives the DTM recovery complete message (mentioned in sequence diagram), will update this explicitly in `DTM Complete` section.

>What will happen if Motr processes die and restart in the middle of DTM recovery? Will recovery proceed automatically? Will the callback be invoked anyways?

Yes, if Motr process restarts during DTM recovery, Hare based on the Failvec and process status takes a decision if the DTM recovery is complete or it needs to re-trigger DTM recovery, have mentioned this in `Concurrent failures` section and sequence diagram.